include/forcing/ForcingsEngineDataProvider.hpp

```diff
@@ -80,50 +61,79 @@ struct ForcingsEngineDataProvider
     
     std::vector<double> get_values(const selection_type& selector, data_access::ReSampleMethod m) override = 0;
 
-  protected:
-    //! Instance map
-    //! @note this map will exist for each of the 
-    //!       3 instance types (lumped, gridded, mesh).
-    static std::unordered_map<
-        std::string,
-        std::unique_ptr<ForcingsEngineDataProvider>
-    > instances_;
 
-    static ForcingsEngineDataProvider* get_instance(
-      const std::string& init,
-      std::size_t time_begin,
-      std::size_t time_end
+    /* Friend functions */
+    static ForcingsEngineDataProvider* instance(
+        const std::string& init,
+        std::size_t time_begin,
+        std::size_t time_end
     )
     {
-        std::unique_ptr<ForcingsEngineDataProvider>& inst = instances_.at(init);
-
+        auto& inst = instances_.at(init);
         if (inst != nullptr) {
-          assert(inst->time_begin_ == clock_type::time_point{std::chrono::seconds{time_begin}});
-          assert(inst->time_end_ == clock_type::time_point{std::chrono::seconds{time_begin}});
+            assert(inst->time_begin_ == clock_type::time_point{std::chrono::seconds{time_begin}});
+            assert(inst->time_end_ == clock_type::time_point{std::chrono::seconds{time_end}});
         }
 
         return inst.get();
     }
 
-    template<typename Callable>
-    static ForcingsEngineDataProvider* set_instance(const std::string& init, Callable&& f)
+  protected:
+
+    // TODO: It may make more sense to have time_begin_seconds and time_end_seconds coalesced into
+    //       a single argument: `clock_type::duration time_duration`, since the forcings engine
+    //       manages time via a duration rather than time points. !! Need to double check
+    ForcingsEngineDataProvider(
+      const std::string& init,
+      std::size_t time_begin_seconds,
+      std::size_t time_end_seconds
+    )
+      : time_begin_(std::chrono::seconds{time_begin_seconds})
+      , time_end_(std::chrono::seconds{time_end_seconds})
     {
-        std::unique_ptr<ForcingsEngineDataProvider> tmp = nullptr;
-        if (instances_[init] != nullptr) {
-            tmp = std::move(instances_[init]);
-        }
-      
-        instances_[init] = f(init);
-        return instances_[init].get();
+        bmi_ = std::make_unique<models::bmi::Bmi_Py_Adapter>(
+            "ForcingsEngine",
+            init,
+            "NextGen_Forcings_Engine.BMIForcingsEngine",
+            /*allow_exceed_end=*/true,
+            /*has_fixed_time_step=*/true,
+            utils::getStdOut()
+        );
+
+        time_step_ = std::chrono::seconds{static_cast<int64_t>(bmi_->GetTimeStep())};
+        var_output_names_ = bmi_->GetOutputVarNames();
     }
 
+    static ForcingsEngineDataProvider* set_instance(
+        const std::string& init,
+        std::unique_ptr<ForcingsEngineDataProvider>&& instance
+    )
+    {
+          using std::swap;
+          swap(instances_[init], instance);
```

Comment:
I'm still not seeing why this needs to me more complicated than
```suggestion
          instances_[init] = std::move(instance);
```
Is there some subtle issue around destructor timing that I'm missing? That's the only reason I can think of for grabbing the existing value in `instances_[init]`

Reply:
No you're right 🤦🏾 This was an artifact of some in-between changes, but after resolving that I forgot to change the swap back to a move assignment